enhancements/update/synchronized-upgrades.md

```diff
@@ -0,0 +1,223 @@
+
+### Non-Goals
+
+- Supporting synchronized upgrades of more than 2 clusters at once.
+
+## Proposal
+
+### User Stories
+
+As the administrator of a cluster using DPUs, I want to be able to do
+z-stream upgrades without causing unnecessary network outages.
+
+### API Extensions
+
+TBD
+
+### Implementation Details/Notes/Constraints [optional]
+
+TBD
+
+### Risks and Mitigations
+
+TBD
+
+## Design Details
+
+### Open Questions
+
+Basically everything...
+
+The general idea is:
+
+- We can set some things up at install time (eg, creating credentials
+  to allow certain operators in the two clusters to talk to each
+  other).
+
+- As part of the DPU security model, the tenant cluster cannot have
+  any power over the infra cluster. (In particular, it can't be
+  possible for an administrator in the tenant cluster to force the
+  infra cluster to upgrade/downgrade to any particular version.) Thus,
+  the upgrade must be initiated on the infra cluster side, and the
+  infra side will tell the tenant cluster to upgrade as well.
+  (Alternatively, the upgrade must be simultaneous-ish-ly initiated in
+  both clusters, if we don't want the infra cluster to have to have a
+  credential that lets it initiate an upgrade in the tenant cluster.)
+
+- An upgrade should not be able to start unless both clusters are able
+  to upgrade.
+
+  - In particular:
+
+    - There can be no `Upgradeable: False` operators in either cluster.
+
+    - The version to upgrade to must be available to both clusters
+      (ie, it must be available for both x86 and ARM).
+
+  - This could be implemented via some sort of "dpu-cluster-upgrade"
+    operator running in both clusters, where the two operators
+    communicate with each other and set their `Upgradeable` state to
+    reflect the state of the other cluster. If the
+    "dpu-cluster-upgrade" operator was placed before every other
+    operator in upgrade priority, then it could also block disallowed
+    upgrades, by failing its own upgrade, eg if an admin tries to
+    upgrade one cluster without the other, or tries to upgrade the two
+    clusters to different versions.
+
+    - (Or should it be possible to do z-stream upgrades of the tenant
+      cluster without bothering to upgrade the infra cluster too?)
+
+- The two clusters upgrade all of the operators up to MCO in parallel.
+
+- Whichever cluster reaches the MCO upgrade first needs to wait for
+  the other cluster to get there before proceeding. The two MCOs then
+  need to coordinate to complete the upgrade: First, they have to
+  agree on what order the physical hosts will be upgraded in. Second,
+  for each physical host, they have to properly synchronize the
+  upgrades of its infra node and its tenant node.
+
+  - More specifically, for each physical host, in some order:
+
+    - The Infra MCO will cordon and drain that host's infra node, and
+      the Tenant MCO will cordon and drain that host's tenant node.
+      (This can happen in parallel.)
```

Comment:
When this happens in parallel, will one MCO waits for the other until it completes the drain, then coordinate on the rebooting? especially the Infra MCO needs to wait until Tenant MCO to finish drain, otherwise infra MCO would reboot DPU which disconnect Tenant api network.

I don't see we have such wait mechanism in MCO.

Reply:
> I don't see we have such wait mechanism in MCO.

Exactly. That's why we need this enhancement.